Objects/moduleobject.c

```diff
@@ -819,68 +872,109 @@ _Py_module_getattro_impl(PyModuleObject *m, PyObject *name, int suppress)
         Py_DECREF(getattr);
         return result;
     }
+
+    if (suppress == 1) {
+        return NULL;
+    }
     if (PyDict_GetItemRef(m->md_dict, &_Py_ID(__name__), &mod_name) < 0) {
         return NULL;
     }
-    if (mod_name && PyUnicode_Check(mod_name)) {
-        PyObject *spec;
-        if (PyDict_GetItemRef(m->md_dict, &_Py_ID(__spec__), &spec) < 0) {
+    if (!mod_name || !PyUnicode_Check(mod_name)) {
+        Py_XDECREF(mod_name);
+        PyErr_Format(PyExc_AttributeError,
+                    "module has no attribute '%U'", name);
+        return NULL;
+    }
+    PyObject *spec;
+    if (PyDict_GetItemRef(m->md_dict, &_Py_ID(__spec__), &spec) < 0) {
+        Py_DECREF(mod_name);
+        return NULL;
+    }
+    PyObject *origin = NULL;
+    if (spec) {
+        int rc = PyObject_GetOptionalAttr(spec, &_Py_ID(origin), &origin);
+        if (rc == -1) {
+            Py_DECREF(spec);
             Py_DECREF(mod_name);
             return NULL;
         }
-        if (suppress != 1) {
-            int rc = _PyModuleSpec_IsInitializing(spec);
+        if (rc == 1 && !PyUnicode_Check(origin)) {
+            Py_DECREF(origin);
+            origin = NULL;
+        }
+    }
```

Comment:
If `spec` is NULL then there isn't much else to do, is there?  `origin` will be NULL, which means `is_possibly_shadowing` will be false, as will `is_possibly_shadowing_stdlib`.

(Furthermore, it *looks* like the later code in this function [^1] expects `spec` to be non-NULL.  I isn't *actually* a problem, but it's always nice when the code looks like it actually behaves.)

[^1]: Specifically the branch at line 930, starting with `rc = _PyModuleSpec_IsUninitializedSubmodule(spec, name);`.

Overall, it seems like we should bail out here.  We could do it as an "else" clause, but I suppose the code is cleaner if we handle the `NULL` case first:

```C
    if (spec == NULL) {
        PyErr_Format(PyExc_AttributeError,
                     "module has no attribute '%U'", name);
        Py_DECREF(mod_name);
        return NULL;
    }

    PyObject *origin = NULL;
    int rc = PyObject_GetOptionalAttr(spec, &_Py_ID(origin), &origin);
    if (rc == -1) {
        Py_DECREF(spec);
        Py_DECREF(mod_name);
        return NULL;
    }
    if (rc == 1 && !PyUnicode_Check(origin)) {
        Py_DECREF(origin);
        origin = NULL;
    } 
```

Reply:
Great point, thank you for the close review! I've pushed changes corresponding to the control flow related feedback, now looking at the other comments